src/main/java/org/assertj/core/api/SoftProxies.java

```diff
@@ -83,58 +106,78 @@ void collectError(Throwable error) {
   }
 
   // TODO V extends AbstractAssert ?
-  @SuppressWarnings("unchecked")
-  <V, T> V create(final Class<V> assertClass, Class<T> actualClass, T actual) {
-
+  <V, T> V create(Class<V> assertClass, Class<T> actualClass, T actual) {
     try {
-      ClassLoader classLoader = getClass().getClassLoader();
-      SimpleKey key = new SimpleKey(assertClass);
-      Class<V> proxyClass = (Class<V>) cache.findOrInsert(classLoader, key, () -> createProxy(assertClass, collector));
-
+      Class<? extends V> proxyClass = createProxy(assertClass);
       Constructor<? extends V> constructor = proxyClass.getConstructor(actualClass);
-      return constructor.newInstance(actual);
+      V instance = constructor.newInstance(actual);
+      ((AssertJProxySetup) instance).assertj$setup(new ProxifyMethodChangingTheObjectUnderTest(this), collector);
```

Comment:
I don't get this line, how come `instance ` is a `AssertJProxySetup` ? I guess the created proxy class actually implements `AssertJProxySetup` but it feels like (black) magic to me.

Reply:
ok this is explained by line 131 `.implement(AssertJProxySetup.class)` later on.

Reply:
This can be done due to all proxies implementing the interface.